File: Atcoder_#292/ruby/c_four_variables.rb
n = gets.to_i
count = 0

# A * B の組み合わせを先に計算
ab_combinations = {}

(1..n).each do |a|
  (1..n).each do |b|
    ab = a * b
    break if ab > n
    ab_combinations[ab] ||= 0
    ab_combinations[ab] += 1
  end
end

# 残りの C * D のペアを探す
(1..n).each do |c|
  (1..n).each do |d|
    cd = c * d
    break if cd > n
    count += ab_combinations[n - cd] if ab_combinations[n - cd]
  end
end

puts count

File: Atcoder_#292/javascript/c_four_variables.js
function main(input) {
    const n = parseInt(input.trim());
    let count = 0;

    // A * B の組み合わせを先に計算
    const abCombinations = new Map();

    for (let a = 1; a <= n; a++) {
        for (let b = 1; b <= n; b++) {
            const ab = a * b;
            if (ab > n) break;
            if (!abCombinations.has(ab)) {
                abCombinations.set(ab, 0);
            }
            abCombinations.set(ab, abCombinations.get(ab) + 1);
        }
    }

    // 残りの C * D のペアを探す
    for (let c = 1; c <= n; c++) {
        for (let d = 1; d <= n; d++) {
            const cd = c * d;
            if (cd > n) break;
            const target = n - cd;
            if (abCombinations.has(target)) {
                count += abCombinations.get(target);
            }
        }
    }

    console.log(count);
}

// 標準入力を読み込んで main 関数を呼び出す
const input = require('fs').readFileSync('/dev/stdin', 'utf8');
main(input);
